WordPress/gutenberg
# Add link color support to Nav block to allow for override of Theme JSON styles

## What?
Fixes https://github.com/WordPress/gutenberg/issues/41146 by adding `linkColor` support to the Nav block thereby ensuring that Nav block specific link color styles can take precedence over inherited Theme JSON styles. 

## Why?


In https://github.com/WordPress/gutenberg/issues/41146#issuecomment-1132775573 we learnt that the Theme JSON styles coming from _link_ color styles on the enclosing Group block (the one wrapping the Nav block) were taking precedence over the user selected text color styles on the Nav block.

This is correct (if unintuative) behaviour as CSS cascade means the link selector from the Global Styles will target the `<a>` of the Nav block. That overrides the text color set on the Nav block.

This is becayse textColor !== linkColor.

We need to provide a way for the Nav block to overide linkColor styles cascading from Theme JSON.

## How?


This PR:

- manually (_not_ via block supports) adds `linkColor` and `customLinkColor` support to the Navigation block.
- ensures the values of the above are passed via context and consumed by 
    - Navigation Submenu block
    - Navigation Link block
- adds appropriate classes to the aforesaid blocks
- ensures that CSS rules and specificity is adjust to ensure that the link color is `inherit`ed from the appropriate parent on to the appropriate element.

The above requires jumping through many hoops and thus why there are so many file changes.


    

## Alternatives considered

Initially this PR targetted the existing `textColor` on the Nav block and ensured that took precedence on any styles cascading from Global Styles. However as @scruffian [pointed out that is counterintuitive](https://github.com/WordPress/gutenberg/pull/41199#issuecomment-1132841240) as link and text color are not the same thing.

One concern here is that for a while Nav block consumers have probably seen the `Text` color control as equivalent to "the color of the links in the Nav". However in the future we may have text items in a Nav which are not links and so we need to start to differentiate them.



## Testing Instructions





- Add a Nav block and some items.
- Wrap the Nav in a Group and set a link color on that group.
- See that by default the Nav inherits the link color set on the Group
- Save and check the front end also respects the above.
- Now set a specific Link Color on the Nav block which is different to that on the Nav block.
- See that this color is respected.
- Save and check the front end also respects the above.
- Now set an `Overlay and submenu link` color on the Nav block. Make sure this distinct from the previous color and the one on the Group block. 
- Check this renders correctly in Editor and front end.
- Remove the link color from the Nav block, whilst retaining the submenu link color. Again check things inherit correctly.
- Try other variations of the above to ensure CSS cascades correctly.
- Check the overlay also obeys these rules.

🤔 We should also test that adding styles from Theme JSON that _specifically_ target the Nav block are still respected.

## Screenshots or screencast 

https://user-images.githubusercontent.com/444434/169521486-8d1a8202-52fc-4198-83de-a14c1604c8e4.mp4
